lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -649,37 +666,39 @@ def do_3d_projection(self, renderer):
             self.update_scalarmappable()
             self._facecolors3d = self._facecolors
 
-        txs, tys, tzs = proj3d._proj_transform_vec(self._vec, renderer.M)
-        xyzlist = [(txs[sl], tys[sl], tzs[sl]) for sl in self._segslices]
+        psegments = proj3d._proj_transform_vectors(self._segments, renderer.M)
+        num_faces = len(psegments)
 
         # This extra fuss is to re-order face / edge colors
         cface = self._facecolors3d
         cedge = self._edgecolors3d
-        if len(cface) != len(xyzlist):
-            cface = cface.repeat(len(xyzlist), axis=0)
-        if len(cedge) != len(xyzlist):
+        if len(cface) != num_faces:
+            cface = cface.repeat(num_faces, axis=0)
+        if len(cedge) != num_faces:
             if len(cedge) == 0:
                 cedge = cface
             else:
-                cedge = cedge.repeat(len(xyzlist), axis=0)
+                cedge = cedge.repeat(num_faces, axis=0)
 
-        # sort by depth (furthest drawn first)
-        z_segments_2d = sorted(
-            ((self._zsortfunc(zs), np.column_stack([xs, ys]), fc, ec, idx)
-             for idx, ((xs, ys, zs), fc, ec)
-             in enumerate(zip(xyzlist, cface, cedge))),
-            key=lambda x: x[0], reverse=True)
+        face_z = self._zsortfunc(psegments[..., 2], axis=-1)
+        if isinstance(face_z, np.ma.MaskedArray):
+            # NOTE: Unpacking .data is safe here, because every face has to
+            #       contain a valid vertex.
+            face_z = face_z.data
+        face_order = np.argsort(face_z, axis=-1)[::-1]
 
-        segments_2d = [s for z, s, fc, ec, idx in z_segments_2d]
         if self._codes3d is not None:
-            codes = [self._codes3d[idx] for z, s, fc, ec, idx in z_segments_2d]
+            segments_2d = [s.data[~s.mask.all(axis=-1)]
```

Comment:
With my implementation in the constructor, you now have the invariant that `mask.all(axis==-1) == mask[:,0]`, which if you wanted to you could use for a very marginal speedup:
```suggestion
            # this invariant is ensured by the constructor
            assert s.mask.strides[1] == 0 
            # which means that s.mask[:,i] is all the same location in memory, so there's no need to call `all`.
            segments_2d = [s.data[~s.mask[:,0]]
```

Reply:
Or I could put in the assert and use `compressed()`, which avoids the unnecessary mask manipulation!

Reply:
`compressed` is internally more complex that that operation anyway though

Reply:
In fact, `mask.ravel()` is now a lot more expensive than `mask[:,0]`, as it has to create a copy to handle the 0 stride.

Reply:
Ok, here's what I came up with:
```diff
@@ -667,6 +667,7 @@ class Poly3DCollection(PolyCollection):
             self._facecolors3d = self._facecolors 
  
         psegments = proj3d._proj_transform_vectors(self._segments, renderer.M)
+        is_masked = isinstance(psegments, np.ma.MaskedArray)
         num_faces = len(psegments)
  
         # This extra fuss is to re-order face / edge colors
@@ -681,19 +682,24 @@ class Poly3DCollection(PolyCollection):
                 cedge = cedge.repeat(num_faces, axis=0)
  
         face_z = self._zsortfunc(psegments[..., 2], axis=-1)
-        if isinstance(face_z, np.ma.MaskedArray): 
+        if is_masked:
             # NOTE: Unpacking .data is safe here, because every face has to
             #       contain a valid vertex.
             face_z = face_z.data
         face_order = np.argsort(face_z, axis=-1)[::-1]
  
+        segments_2d = psegments[face_order, :, :2]
         if self._codes3d is not None:
-            segments_2d = [s.data[~s.mask.all(axis=-1)]
-                           for s in psegments[face_order, :, :2]]
+            if is_masked:
+                # NOTE: We cannot assert the same on segments_2d, as it is a
+                #       result of advanced indexing, so its mask is a newly
+                #       allocated tensor. However, both of those asserts are
+                #       equivalent.
+                assert psegments.mask.strides[-1] == 0
+                segments_2d = [s.compressed().reshape(-1, 2) for s in segments_2d]
             codes = [self._codes3d[idx] for idx in face_order]
             PolyCollection.set_verts_and_codes(self, segments_2d, codes)
         else:
-            segments_2d = psegments[face_order, :, :2]
             PolyCollection.set_verts(self, segments_2d, self._closed)
  
         self._facecolors2d = cface[face_order]
```

Reply:
Those `.ravel()`s are no-ops, because `segments_2d` comes from advanced indexing anyway, so I still went with `compressed()`.